### ISSUE TITLE
Unable to boot | "qubes_dom0-root does not exist" | Possible LVM problem

### ISSUE THREAD

**Issue by gga10** [QubesOS/qubes-issues#8707]
As requested [here](https://github.com/QubesOS/qubes-issues/issues/5973#issuecomment-1806922191), I am opening a new issue for this bug.

### Qubes OS release
R4.1.2

### Affected component(s) or functionality
Entire system

### Brief summary
After months of using Qubes, system is unable to start after disk decryption. 

### Steps to reproduce
Unknown
Uneducated guess to reproduce the bug (based on related previously reported issues): run out of disk space

### Expected behavior
Normal system startup, with the login screen appearing after disk decryption.

### Actual behavior
The following error screen appears after I enter the password to decrypt the disk:

![01](https://github.com/QubesOS/qubes-issues/assets/150536125/210d57e9-6e94-477f-a3a4-75412998950e)

Here is the generated [rdsosreport.txt](https://github.com/QubesOS/qubes-issues/files/13329223/rdsosreport.txt)

If I press Control-D to continue, this is what I get:
![02](https://github.com/QubesOS/qubes-issues/assets/150536125/fba25b59-bc4e-438d-845e-fa48800deb9b)

### Additional context
- My keyboard strangely stopped working out of nowhere the session before the bug happened;
- Still in the session before and in other previous (but recent) instances, system prompted a warning message stating that I was running out of space. I don't recall exactly, but at the time I assumed that the space it was referring to was one of my VMs, since that warning message only appeared after starting that VM, so I simply increased the space allocated to it in its settings;
- After this bug started happening, the decryption process is taking a much longer time than before just to take me to that error screen. However, one thing I have certainly noticed is that the time to decrypt the disk has gradually become longer over time since Qubes fresh install.
- I managed to recover the files that were in the VMs to an external drive by booting from a live usb Linux Mint image. After starting Mint, the drive in which Qubes is installed appeared as available. I just double clicked it, entered the password to decrypt the disk, and voila, the VMs showed up.

### Solutions you've tried
After inspecting rdsosreport.txt (journalctl), I noticed a familiar “Manual repair required!” warning, something common in the other reported cases:
> [   40.070355] dom0 dracut-initqueue[1111]: Check of pool qubes_dom0/root-pool failed (status:1). Manual repair required!
[  181.207297] dom0 dracut-initqueue[554]: Warning: dracut-initqueue timeout - starting timeout scripts

So I tried to follow the solutions suggested in these other cases.

In the maintenance/emergency shell, I tried running the command `lmv lvconvert --repair qubes_dom0/root-pool`, which resulted in the same “Manual repair required!” error:
![03](https://github.com/QubesOS/qubes-issues/assets/150536125/c619cfef-ca9d-4dad-98d6-95005767413d)

Following the assumptions that this problem was related to the lack of disk space, I deleted one of my VMs that had a large space allocated using the `lvremove` command. I was successful in removing the VMs, but the problem persisted.

One of the helpful community members theorized that I ran out of space in dom0 (root-pool) and suggested me to free up the space in it by extending the volume. So I tried to extend it, but I was unsuccessful.
![04](https://github.com/QubesOS/qubes-issues/assets/150536125/fe818b0c-ec6b-4798-b170-0cc03838927c)

I also misinterpreted one of his instructions, and ended up trying to extend the vm-pool as well, despite its lack of correlation with the root-pool. Still, I came across this rather questionable warning message that might be worth mentioning:
![05](https://github.com/QubesOS/qubes-issues/assets/150536125/e331de2a-fe66-4c61-a245-003c0c817a5f)

### Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues
https://github.com/QubesOS/qubes-issues/issues/5372
https://github.com/QubesOS/qubes-issues/issues/5973
https://www.reddit.com/r/Qubes/comments/fszsu7/qubes_suddenly_broken_warning_devqubes_dom0root/?rdt=61290
https://forum.qubes-os.org/t/qubes-dom0-root-does-not-exist/8600/23
https://forum.qubes-os.org/t/emergency-dracut-shell-qubes-not-working-for-no-reason/6717/3

**Comment by DemiMarie** [comment]
“bad checksum in superblock” looks like disk corruption.

**Comment by gga10** [comment]
> “bad checksum in superblock” looks like disk corruption.

After a long test, the SMART of the drive showed no problems except for its temperature, which is odd because the temperature was sitting at 50-52°C and reached 60°C during test. Not the best result, but far from being critical. But yes, I must say that this is a relatively old drive.

**Comment by jamke** [comment]
> “bad checksum in superblock” looks like disk corruption.

Why? The same would happen if data of the disk/LVM is corrupted, by software.

**Was this situation (when LVM is out of space) ever tested?** I think if somebody has spare Qubes OS installation (R4.1 preferably) than exhausting all the space of LVM and working in this state for a day may reproduce this issue. And the consequences of being out of space in Qubes OS may be huge, including user data loss and unbootable system.

---

Two comments about the same situation after being out of space:

>On R3.2, or R4.0 I had very similar situation. I **ran out of space**, something bad was happening with the system after that. So, I rebooted and found out that **LVM was broken**, I managed to extract only part of the the information from the LVM using LiveCD because **LVM was indeed corrupted**.
>
>All attempts to fix it in place failed and I had to reinstall the system completely. The **SMART of the drive showed no problems**.

And:

>I booted Ubuntu LiveCD and after some struggle was able to manually mount private LVM volumes of qubes and extract part of the information. Unfortunately not all, but I had up to date backup. I tried different ways of recovering LVM from the internet, but they all failed.
>
>**After that I do not trust LVM that much and keep 50%+ of free space on the Qubes OS disk**, because I fear that it can collapse unrecoverable when it runs out of space.
>
>... I decided not to report it because I did not find the same cases, so I considered the possibility that the reason is random hardware failure or something rare like cosmic rays that happened the day when I ran out of space just by chance. And now I do not recall details for the report.

**Comment by DemiMarie** [comment]
@gga10 >I still think that this is a disk problem.  Even when the disk fills up, the thin pool _itself_ should still have consistent metadata, and yours is not consistent.

Can you run `lvconvert --repair --verbose`?  That might indicate what `lvm` is doing under the hood.  It should be calling `thin_repair` to try to repair the metadata.</summary>

<details><summary>Technical details of why there have been problems in the past.</summary>

There are two problems that make out of space recovery hard:

1. LVM is optimized for servers, not desktops.
2. Qubes OS exposes storage at the block level, not the file level.

LVM’s full-time developers work for Red Hat, whose customers use LVM in enterprise servers.  Therefore, LVM is optimized for enterprise servers, where avoiding downtime is critical.  Since running out of space in a thin pool means downtime, LVM considers letting a pool run out of space to be administrative error.  One must ensure that the pool does not fill up in the first place by adding storage, deleting snapshots, etc.  LVM makes this easier by automatically extending pools that run low on space.  LVM is much better at recovering from space exhaustion than it used to be, but in the past there were quite a few bugs here.  My understanding is that these have been fixed, though, which is why I am so surprised at the problem you are having.

If a pool does fill up, the result is that the filesystem gets ENOSPC when it tries to write back dirty data.  In the past, filesystems handled this very badly and would become corrupted.  Nowadays, filesystems are generally better at not becoming corrupted, but that is all.  Furthermore, the supported methods for recovering from a full thin pool are to either grow the pool or to delete thin volumes.  This is because trying to mount a filesystem may require writing to it, and that might cause ENOSPC all over again.

The proper fix to (1) is to use a storage stack, such as ZFS or bcachefs, that is actually intended for client machines and has better out of space handling.  ZFS, and I believe bcachefs, both reserve a certain amount of space (I think 10% or so) to ensure that they can recover from full devices.  The proper fix for (2) is to expose storage at the filesystem level, not the block device level, or to at least ensure that the guest is aware that the underlying storage is thinly provisioned and knows what operations are and are not guaranteed to work. 

Unfortunately, neither of these solutions are easy.  BTRFS has performance problems, ZFS has licensing problems, and bcachefs is still experimental.  virtiofs has much more attack surface than blkback, and I am not aware of any plans to make blkfront/blkback thin provisioning aware.
</details>

**Comment by jamke** [comment]
@DemiMarie thank you for technical details! Long story short, what exactly happens nowadays when user runs out of space in Qubes OS R4.2? Is it still that bad (gets unbootable, can still destroy data on ext4)? Or was it partly solved by reorganizing `dom0` partitions recently (after R4.0?)?

**Comment by DemiMarie** [comment]
@jamke It should not destroy any data anywhere (if it does, that’s a bug), but it might be difficult to recover from without deleting something (such an old revision of a VM) to free up space.

**Comment by jamke** [comment]
@So, it still becomes unbootable and require manual recovery that is even hard for advanced users with spare computer? It does not look good.

`dom0` seems to be in a different separated pool nowadays. Doesn't it mean, that `dom0` should stay completely working and bootable with no issues (with qubes autostart disabled)?

**Comment by marmarek** [comment]
Yes, dom0 will remain fully functional, but if vm-pool is full, no VM
can start.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab

**Comment by gga10** [comment]
@jamke I'm glad you mentioned ``qubes autostart disabled`` because I actually tried that and forgot to mention it when opening this issue. It had no effect, unfortunately.

@DemiMarie Here is the result for ``lvconvert --repair --verbose``

![InShot_20231115_175502764](https://github.com/QubesOS/qubes-issues/assets/150536125/b80954c1-fcd3-4291-a06c-93c62fb6373c)

It is also interesting how you mentioned that LVM is optimized for servers. In this case, is there any difference in how LVM works depending on the hardware on which the system was installed? I ask this because I was actually running Qubes on a server like setup. It has a Xeon E5-2690 V4, x99 mobo and ECC RAM.

So I might be able to recover the system if I delete something, like an old revision of a VM. How do you suggest me to do that?

**Comment by DemiMarie** [comment]
@gga10 Can you see if `thin_dump` on the device (`/dev/mapper/qubes_dom0-root--pool_tmeta` in your case) with the corrupted metadata returns anything?

“bad checksum in superblock” means that the superblock of the metadata is corrupt, and it looks like `thin_repair` is refusing to repair it.  I recommend reaching out on the linux-lvm mailing list for help, but is not something that is guaranteed to be repairable.

> It is also interesting how you mentioned that LVM is optimized for servers. In this case, is there any difference in how LVM works depending on the hardware on which the system was installed? I ask this because I was actually running Qubes on a server like setup. It has a Xeon E5-2690 V4, x99 mobo and ECC RAM.

I don’t think so, as you found out.

**Comment by jamke** [comment]
Can somebody actually check what happens when Qubes OS R4.1 runs out of space? Does it corrupt LVM or not?

Maybe it does after some attempts to write some extra data with qvm-copy or something.
Because in my case it also became unbootable (with and without qubes autostart). Corrupted and I failed to repair.

I thought it was a random coincidence with running out of space the same day, but now, after the second case, I changed my mind and leaning towards Qubes OS destroying itself when the thin pool gets out of space.

**Comment by gga10** [comment]
> Can you see if `thin_dump` on the device (`/dev/mapper/qubes_dom0-root--pool_tmeta` in your case) with the corrupted metadata returns anything?

If I'm not missing anything, I'm afraid I can't unfortunately. As you can see in my last screenshot, `/dev/mapper/qubes_dom0-root--pool_tmeta` is automatically removed at the end.

**Comment by dhhjdhhhhfhfcnncnnn** [comment]
This randomly happened to me on Qubes 4.2.4. I was using the PC normally when all my Qubes (except for a ProtonVPN qube I had set up) froze and I wasn't able to close any of them. Then they closed themselves and  eventually the VPN Qube froze up, the text outside the VPN Qube started showing up as blocks, the top navigation bar disappeared (I was still able to open the context menu with a right click for a bit, but it stopped opening after I hovered over one of the options), and eventually Qubes OS turned itself off.

My drive is only a few years old and hasn't shown any issues in smartctl. I've never had data corruption issues with this drive before. I have a fairly recent backup and I will try restoring this tomorrow but I have never had this occur on any other Linux distribution and in my opinion this seems like a Qubes bug (how would dom0 simply stop existing? what sudden drive corruption issue would occur with no prior issues on a genuine drive?)

I need to get to sleep, so I will post more information on my system and the error messages I have in the morning. I took these pictures with my phone camera, and I'll add more when I can.

![Image](https://github.com/user-attachments/assets/39ed5651-c350-4a12-9ebd-03fe6843ed51)
![Image](https://github.com/user-attachments/assets/16a7b846-75dd-4fb0-af49-28e5900100e9)

**Comment by dhhjdhhhhfhfcnncnnn** [comment]
I just started it up after waiting overnight and got to the desktop, but the start menu at the top right froze up after I clicked on it. All the buttons on the screen stopped being responsive, and my VPN Qube that automatically starts on boot closed itself, the navigation bar at the top disappeared after a minute or so, and the context menu doesn't open when I right click. My USB mouse was working fine for a little bit, but then stopped working, so I had to use the trackpad. My external USB hard drive gets power and turns its light on when I plug it in, but nothing happens in QubesOS as far as I can see to indicate it recognizes the device. After the navigation bar disappeared, the context menu works again but shows the weird boxes.

![Image](https://github.com/user-attachments/assets/89deac36-bc18-4c42-ae1e-0ca423e3405b)

I also recorded a video, but GitHub won't let me upload it. Basically, it shows that hovering over the last item in the context menu causes a dialogue to pop up that won't be closed even if I press the button on it, press esc, press alt +f4, etc. Eventually the screen turns to black but the PC doesn't turn off. The fans slowed down after a minute or so.

**Comment by dhhjdhhhhfhfcnncnnn** [comment]
Tried restarting my PC after that. Black screen for a while, eventually the fans sped up but didn't get past the black screen. I saw some text pop up on screen when I pressed random keys, but it disappeared before I could read it. So I rebooted and got to the UEFI system test.

The system test in the UEFI states that I do not have a hard drive installed after that happened. It said the same thing yesterday. You would think this means the drive is bad, but it stops saying that and boots right back into Qubes if I just turn the laptop off, remove the battery, and wait a few seconds before putting it back in and restarting, or sometimes even just try again by restarting the PC without doing anything. When I get back into Qubes after that, it will proceed to either be weird as described previously or just say that dom0 doesn't exist. Qubes will usually but not always take a ridiculously long time to decrypt the drive.

Also, when I look inside the laptop, the SSD seems to be connected and there are no signs of overheating or clogged fans, so I doubt it would suddenly fail without having any previous issues. I also never had any drive issues on Windows, Ubuntu, or Arch Linux, all of which I have installed on this laptop.

Currently I am back to the screen saying dom0 doesn't exist.

![Image](https://github.com/user-attachments/assets/140aee1a-9132-4d0f-b02a-41eea56f6c1c)

**Comment by gga10** [comment]
Hello! I've just seen that this issue has been reopened.

I would like to say that I no longer use Qubes on the same disk. Shortly after this problem happened, I've decided to go back to basics and install Linux Mint on the drive. I haven't experienced any disk problems since then.

The only strange thing I remember happening was right at the moment of Mint installation process, which was crashing constantly. As a solution, I first had to fully format the disk with GParted and then install Mint with LVM using the option to overwrite empty disk space. Any configuration other than that would cause the installation to fail again.

My line of thought was that if there were corrupted files that were preventing the installation of the new system, overwriting them could solve the problem. And indeed, that was the solution. I honestly don't know if this reasoning makes any sense technically speaking, nor if it has anything to do with the original problem, but that's what happened.

I hope you find a solution my friend @dhhjdhhhhfhfcnncnnn After all, I myself plan to buy a new notebook and return to using Qubes with it. I hope the cause of this problem can be discovered by then!
